modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/raft/MetaStorageSnapshotStorageFactory.java

```diff
@@ -0,0 +1,109 @@
+import org.apache.ignite.raft.jraft.storage.snapshot.SnapshotReader;
+import org.apache.ignite.raft.jraft.storage.snapshot.local.LocalSnapshotStorage;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Snapshot storage factory implementation for Meta Storage.
+ */
+public class MetaStorageSnapshotStorageFactory implements SnapshotStorageFactory {
+    private final KeyValueStorage storage;
+
+    /** Snapshot meta, constructed from the storage data and raft group configuration at startup. {@code null} if the storage is empty. */
+    private final @Nullable RaftOutter.SnapshotMeta startupSnapshotMeta;
+
+    /**
+     * Constructor.
+     *
+     * @param storage Key-value storage instance.
+     */
+    public MetaStorageSnapshotStorageFactory(KeyValueStorage storage) {
+        this.storage = storage;
+
+        startupSnapshotMeta = readStartupSnapshotMeta();
+    }
+
+    private @Nullable SnapshotMeta readStartupSnapshotMeta() {
+        IndexWithTerm indexWithTerm = storage.getIndexWithTerm();
+
+        if (indexWithTerm == null) {
+            return null;
+        }
+
+        byte[] configBytes = storage.getConfiguration();
+        assert configBytes != null;
+
+        RaftGroupConfiguration configuration = ByteUtils.fromBytes(configBytes);
+        assert configuration != null;
+
+        return new RaftMessagesFactory().snapshotMeta()
+                .lastIncludedIndex(indexWithTerm.index())
+                .lastIncludedTerm(indexWithTerm.term())
+                .peersList(configuration.peers())
+                .oldPeersList(configuration.oldPeers())
+                .learnersList(configuration.learners())
+                .oldLearnersList(configuration.oldLearners())
+                .build();
+    }
+
+    /**
+     * Returns a snapshot meta, constructed from the storage data and raft group configuration at startup.
+     */
+    SnapshotMeta startupSnapshotMeta() {
+        if (startupSnapshotMeta == null) {
+            throw new IllegalStateException("Storage is empty, so startup snapshot should not be read");
+        }
+
+        return startupSnapshotMeta;
+    }
+
+    @Override
+    public @Nullable SnapshotStorage createSnapshotStorage(String uri, RaftOptions raftOptions) {
+        return new LocalSnapshotStorage(uri, raftOptions) {
+            private final AtomicBoolean startupSnapshotOpened = new AtomicBoolean(false);
+
+            @Override
+            public SnapshotReader open() {
+                if (startupSnapshotOpened.compareAndSet(false, true)) {
+                    if (startupSnapshotMeta == null) {
+                        // The storage is empty, let's behave how JRaft does: return null, avoiding an attempt to load a snapshot
+                        // when it's not there.
+                        return null;
+                    }
+
+                    return new StartupMetaStorageSnapshotReader(MetaStorageSnapshotStorageFactory.this);
```

Comment:
This is so weird. Why don't we pass the `startupSnapshotMeta` here and instead have this circular reference?

Reply:
Even more, I would expect the `StartupMetaStorageSnapshotReader` to read the startup meta and instead it is done by the factory for some reason

Reply:
>  read the startup meta and instead it is done by the factory for some reason

I have to read meta in order to determine whether I need to return `null` or `StartupMetaStorageSnapshotReader`. When it comes to circular references, my bad